### PR TITLE
Support local cluster edge case testing

### LINKED ISSUE #7086: Use sleep_n_epochs instead of sleep for 60 seconds
Is the goal here to start the partition after all the validators are already in the leader schedule? If so, there's some logic in `cluster_tests::sleep_n_epochs` that can more precisely calculate how long to wait based on epoch len, ticks per slot, etc.

_Originally posted by @carllin in https://github.com/solana-labs/solana/pull/7079_

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Local cluster is missing the ability to set up tests for edge cases

#### Summary of Changes
1) Add ability to set up a specific leader schedule to test specific scenarios
2) Add ability to kill a partition in local cluster
3) Add test that tests killing a partition when
     a) It is the majority partition
     b) The rest of the network is locked out on the dead, majority partition

Fixes #7086

### CODE REVIEW

File: local-cluster/src/cluster_tests.rs [hunk 1]
```diff
@@ -169,6 +172,11 @@ pub fn verify_ledger_ticks(ledger_path: &Path, ticks_per_slot: usize) {
     }
 }
 
+pub fn time_untl_nth_epoch(epoch: u64, slots_per_epoch: u64, stakers_slot_offset: u64) -> u64 {
```
Comment (garious) [inline local-cluster/src/cluster_tests.rs]: untl?

File: local-cluster/src/local_cluster.rs [hunk 2]
```diff
@@ -129,9 +131,20 @@ impl LocalCluster {
 
     pub fn new(config: &ClusterConfig) -> Self {
         assert_eq!(config.validator_configs.len(), config.node_stakes.len());
-        let leader_keypair = Arc::new(Keypair::new());
+        let validator_keys = {
+            if let Some(ref keys) = config.validator_keys {
+                assert_eq!(config.validator_configs.len(), keys.len());
+                keys.clone()
+            } else {
+                (0..config.validator_configs.len())
```
Comment (garious) [inline local-cluster/src/local_cluster.rs]: fyi, `std::iter::repeat_with`

```rust
iter::repeat_with(|| Arc::new(Keypair::new())).take(config.validator_configs.len()).collect()
```

File: local-cluster/src/local_cluster.rs [hunk 3]
```diff
@@ -129,9 +131,20 @@ impl LocalCluster {
 
     pub fn new(config: &ClusterConfig) -> Self {
         assert_eq!(config.validator_configs.len(), config.node_stakes.len());
-        let leader_keypair = Arc::new(Keypair::new());
+        let validator_keys = {
+            if let Some(ref keys) = config.validator_keys {
```
Comment (garious) [inline local-cluster/src/local_cluster.rs]: As far as I know, you should never need `ref` annotations anymore.

File: local-cluster/tests/local_cluster.rs [hunk 4]
```diff
@@ -188,71 +193,157 @@ fn test_leader_failure_4() {
     );
 }
 
-fn run_network_partition(partitions: &[usize]) {
+fn run_network_partition(
```
Comment (garious) [inline local-cluster/tests/local_cluster.rs]: `network` -> `cluster`.  especially in the module called `local_cluster`

File: local-cluster/tests/local_cluster.rs [hunk 5]
```diff
@@ -188,71 +193,157 @@ fn test_leader_failure_4() {
...
+            };
+            validator_config.fixed_leader_schedule = Some(fixed_schedule);
+            (
+                validator_keys,
+                num_slots_per_rotation * clock::DEFAULT_MS_PER_SLOT,
+            )
+        } else {
+            (
+                (0..partitions.len())
+                    .map(|_| Arc::new(Keypair::new()))
```
Comment (garious) [inline local-cluster/tests/local_cluster.rs]: Instead of `repeat_with` as mentioned above, there's also a parallelized one in `solana_core` called `GenKeys::gen_n_keypairs`.

File: local-cluster/tests/local_cluster.rs [hunk 6]
```diff
@@ -188,71 +193,157 @@ fn test_leader_failure_4() {
     );
 }
 
-fn run_network_partition(partitions: &[usize]) {
+fn run_network_partition(
+    partitions: &[&[(usize, bool)]],
```
Comment (garious) [inline local-cluster/tests/local_cluster.rs]: Need docs for this function.  Nothing obvious about what this type means.

File: local-cluster/src/local_cluster.rs (hunk 3 above)
Comment (carllin) [reply local-cluster/src/local_cluster.rs]: hmmm still complains of a move of a shared reference unless I annotate it here

Comment (garious) [reply local-cluster/src/local_cluster.rs]: weird. thanks for checking
